harness/determined/pytorch/_trainer.py

```diff
@@ -0,0 +1,233 @@
+    # offset in the dataset, ensuring that it is processing a unique
+    # training batch.
+    random.seed(seed)
+    np.random.seed(seed)
+    torch.random.manual_seed(seed)
+
+
+def _generate_local_seed() -> int:
+    return random.randint(0, 1 << 31)
+
+
+@contextlib.contextmanager
+def init(
+    *,
+    hparams: Optional[Dict] = None,
+    exp_conf: Optional[Dict[str, Any]] = None,
+    distributed: Optional[core.DistributedContext] = None
+) -> pytorch.PyTorchTrialContext:
+    cluster_info = det.get_cluster_info()
+    local_training = cluster_info is None or cluster_info.task_type != "TRIAL"
+
+    # Pre-execute steps: initialize distributed backend and random seeds
+    distributed_context = distributed
+
+    if not local_training:
+        distributed_context = _initialize_distributed_backend()
+
+    # Initialize default values
+    if local_training:
+        trial_seed = _generate_local_seed()
+
+        # XXX: figure out if better way to handle this
+        aggregation_frequency = exp_conf and int(exp_conf.get("optimizations", {}).get("aggregation_frequency", 1)) or 1  # type: ignore
+        fp16_compression = False
+        average_aggregated_gradients = True
+        steps_completed = 0
+        managed_training = True
+        debug_enabled = False
+        num_gpus = len(gpu.get_gpu_uuids())
+    else:
+        assert cluster_info, "Unable to detect cluster info"
+
+        trial_seed = cluster_info.trial.trial_seed
+        exp_conf = cluster_info.trial._config
+        aggregation_frequency = int(exp_conf["optimizations"]["aggregation_frequency"])
+        fp16_compression = bool(exp_conf["optimizations"]["gradient_compression"])
+        average_aggregated_gradients = bool(
+            exp_conf["optimizations"]["average_aggregated_gradients"]
+        )
+        steps_completed = cluster_info.trial._steps_completed
+        managed_training = True
+        num_gpus = len(cluster_info.gpu_uuids)
+        debug_enabled = cluster_info.trial._debug
+
+    _set_random_seeds(trial_seed)
+
+    with core.init(
+        distributed=distributed_context,
+        preempt_mode=core.PreemptMode.WorkersAskChief,
+        tensorboard_mode=core.TensorboardMode.MANUAL,
+    ) as core_context:
+        context = pytorch.PyTorchTrialContext(
+            core_context=core_context,
+            trial_seed=trial_seed,
+            hparams=hparams,
+            slots_per_trial=core_context.distributed.get_size(),
+            num_gpus=num_gpus,
+            exp_conf=exp_conf,
+            aggregation_frequency=aggregation_frequency,
+            steps_completed=steps_completed,
+            managed_training=managed_training,
+            debug_enabled=debug_enabled,
+        )
+
+        # Set here for backwards-compatibility: future code-paths should call wrap_optimizer
+        context._set_gradient_compression(fp16_compression)
+        context._set_average_aggregated_gradients(average_aggregated_gradients)
```

Comment:
there shouldn't be any backwards-compatibility in `det.pytorch.init()`, since `det.pytorch.init()` _is_ the codepath future code will go through.

Instead put all this in the actual backwards compatibility layer in `exec/harness.py`.